File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/PmdTimeStampUtils.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import androidx.annotation.VisibleForTesting
import com.polar.androidcommunications.api.ble.exceptions.NegativeTimeStampError
import com.polar.androidcommunications.api.ble.exceptions.SampleSizeMissingError
import com.polar.androidcommunications.api.ble.exceptions.TimeStampAndFrequencyZeroError
import kotlin.math.round

internal object PmdTimeStampUtils {

    fun getTimeStamps(previousFrameTimeStamp: ULong, frameTimeStamp: ULong, samplesSize: Int, sampleRate: Int): List<ULong> {
        // guard
        if (samplesSize <= 0) {
            throw SampleSizeMissingError
        }

        val timeStampDelta = getTimeStampDelta(previousFrameTimeStamp, frameTimeStamp, samplesSize, sampleRate)

        // guard
        if (frameTimeStamp.toDouble() < (timeStampDelta * samplesSize.toDouble())) {
            throw NegativeTimeStampError("Sample time stamp calculation fails. The timestamps are negative, since frameTimeStamp $frameTimeStamp minus ${timeStampDelta * samplesSize} is negative")
        }

        val startTimeStamp = if (previousFrameTimeStamp == 0uL) {
            firstSampleTimeFromSampleRate(frameTimeStamp, timeStampDelta, samplesSize)
        } else {
            firstSampleTimeFromTimeStamps(previousFrameTimeStamp, timeStampDelta)
        }

        val timeStampList = MutableList(size = samplesSize - 1, init = { index ->
            round(startTimeStamp + timeStampDelta * index).toULong()
        })
        timeStampList.add(frameTimeStamp)
        return timeStampList
    }

    private fun getTimeStampDelta(previousFrameTimeStamp: ULong, timeStamp: ULong, samplesSize: Int, sampleRate: Int): Double {
        // guard
        if (previousFrameTimeStamp == 0uL && sampleRate <= 0) {
            throw TimeStampAndFrequencyZeroError("Timestamp delta cannot be calculated for the frame, because previousTimeStamp $previousFrameTimeStamp and sampleRate $sampleRate")
        }

        val delta = if (previousFrameTimeStamp == 0uL) {
            deltaFromSamplingRate(sampleRate)
        } else {
            deltaFromTimeStamps(previousFrameTimeStamp, timeStamp, samplesSize)
        }
        return delta
    }

    @VisibleForTesting
    fun deltaFromSamplingRate(samplingRate: Int): Double {
        return (1.0 / samplingRate.toDouble()) * 1000 * 1000 * 1000
    }

    fun deltaFromTimeStamps(previousTimeStamp: ULong, timeStamp: ULong, samples: Int): Double {
        val timeInBetween = timeStamp - previousTimeStamp
        return if (timeInBetween > 0u) {
            timeInBetween.toDouble() / samples.toDouble()
        } else {
            throw NegativeTimeStampError("Failed to decide delta from when previous timestamp: $previousTimeStamp timestamp: $timeStamp")
        }
    }

    private fun firstSampleTimeFromTimeStamps(previousTimeStamp: ULong, timeStampDelta: Double): Double {
        return previousTimeStamp.toDouble() + timeStampDelta
    }

    private fun firstSampleTimeFromSampleRate(lastSampleTimeStamp: ULong, timeStampDelta: Double, samplesSize: Int): Double {
        return if (samplesSize > 0) {
            val startTimeStamp = lastSampleTimeStamp.toDouble() - (timeStampDelta * (samplesSize - 1))
            if (startTimeStamp > 0) {
                startTimeStamp
            } else {
                throw NegativeTimeStampError("Failed to estimate first sample timestamp when timeStamp: $lastSampleTimeStamp delta: $timeStampDelta size: $samplesSize")
            }
        } else {
            throw NegativeTimeStampError("Failed to estimate first sample timestamp when timeStamp: $lastSampleTimeStamp delta: $timeStampDelta size: $samplesSize")
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/model/PmdTimeStampUtils.swift
//  Copyright © 2022 Polar. All rights reserved.

import Foundation

internal class PmdTimeStampUtils {
    
    static func getTimeStamps(previousFrameTimeStamp: UInt64, frameTimeStamp: UInt64, samplesSize: UInt, sampleRate: UInt) throws -> [UInt64] {
        
        guard samplesSize > 0 else {
            throw BleGattException.gattDataError(description: "Timestamp calculation error, sample size is zero")
        }
        
        let timeStampDelta = try getTimeStampDelta(previousFrameTimeStamp, frameTimeStamp, samplesSize, sampleRate)

        // guard
        guard (Double(frameTimeStamp) >= timeStampDelta * Double(samplesSize)) else {
            throw BleGattException.gattDataError(description: "Sample time stamp calculation fails. The timestamps are negative, since frameTimeStamp \(frameTimeStamp) minus \((UInt64(timeStampDelta) * UInt64(samplesSize))) is negative")
        }

        let startTimeStamp:Double
        if (previousFrameTimeStamp <= 0) {
            startTimeStamp = try firstSampleTimeFromSampleRate(frameTimeStamp, timeStampDelta, samplesSize)
        } else {
            startTimeStamp = firstSampleTimeFromTimeStamps(previousFrameTimeStamp, timeStampDelta)
        }

        var timeStampList = [UInt64]()
        for i in 0..<(samplesSize - 1) {
            timeStampList.append(UInt64(round(startTimeStamp + timeStampDelta * Double(i))))
        }
        timeStampList.append(frameTimeStamp)
        return timeStampList
    }
    
    static func getTimeStampDelta(_ previousFrameTimeStamp: UInt64, _ timeStamp: UInt64, _ samplesSize: UInt, _ sampleRate: UInt) throws -> Double {
        
        guard previousFrameTimeStamp > 0 || sampleRate > 0 else {
            throw BleGattException.gattDataError(description: "Timestamp delta cannot be calculated for the frame, because previousTimeStamp \(previousFrameTimeStamp) and sampleRate \(sampleRate)")
        }
        
        if (previousFrameTimeStamp <= 0) {
            return deltaFromSamplingRate(sampleRate)
        } else {
            return try deltaFromTimeStamps(previousFrameTimeStamp, timeStamp, samplesSize)
        }
    }
    
    internal static func deltaFromSamplingRate(_ samplingRate: UInt) -> Double {
        return (1.0 / Double(samplingRate)) * 1000 * 1000 * 1000
    }
    
    internal static func deltaFromTimeStamps(_ previousTimeStamp: UInt64, _ timeStamp: UInt64, _ samples: UInt) throws -> Double {
        let timeInBetween = timeStamp - previousTimeStamp
        if (timeInBetween > 0) {
            return Double(timeInBetween) / Double(samples)
        } else {
            throw BleGattException.gattDataError(description: "Failed to decide delta from when previous timestamp: \(previousTimeStamp) timestamp: \(timeStamp)")
        }
    }
    
    private static func firstSampleTimeFromSampleRate(_ lastSampleTimeStamp: UInt64, _ timeStampDelta: Double, _ samplesSize: UInt) throws -> Double {
          if (samplesSize > 0) {
             let startTimeStamp = Double(lastSampleTimeStamp) - (timeStampDelta * Double(samplesSize - 1))
             if (startTimeStamp > 0) {
                 return startTimeStamp
             } else {
                 throw BleGattException.gattDataError(description: "Failed to estimate first sample timestamp when timeStamp: \(lastSampleTimeStamp) delta: \(timeStampDelta) size: \(samplesSize)")
             }
         } else {
             throw BleGattException.gattDataError(description:"Failed to estimate first sample timestamp when timeStamp: \(lastSampleTimeStamp) delta: \(timeStampDelta) size: \(samplesSize)")
         }
     }
    
    private static func firstSampleTimeFromTimeStamps(_ previousTimeStamp: UInt64,_ timeStampDelta: Double) -> Double {
         return Double(previousTimeStamp) + timeStampDelta
     }
}
